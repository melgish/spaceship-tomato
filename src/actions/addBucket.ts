'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddBucket, addBucketSchema } from '@/schemas/addBucket.schema'

/**
 * Add server action to a new bucket to the database.
 * @see ../schemas/addBucket.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addBucket(payload: AddBucket) {
  try {
    const data = await addBucketSchema.parseAsync(payload)
    const created = await prisma.bucket.create({
      data: { ...data, plantId: data.plantId || null },
    })
    revalidatePath('/buckets')
    return { ok: true, message: 'Bucket created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create bucket' } as const
  }
}
