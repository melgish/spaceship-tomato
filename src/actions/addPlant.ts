'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddPlant, addPlantSchema } from '@/schemas/addPlant.schema'

/**
 * Add server action to a new plant to the database.
 * @see ../schemas/addPlant.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addPlant(payload: AddPlant) {
  try {
    const data = await addPlantSchema.parseAsync(payload)
    const created = await prisma.plant.create({ data })
    revalidatePath('/plants')
    return { ok: true, message: 'Plant created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create plant' } as const
  }
}
