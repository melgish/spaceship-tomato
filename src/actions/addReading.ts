'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddReading, addReadingSchema } from '@/schemas/addReading.schema'

/**
 * Add server action to a new plant to the database.
 * @see ../schemas/addReading.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addReading(payload: AddReading) {
  try {
    const data = await addReadingSchema.parseAsync(payload)
    const created = await prisma.reading.create({ data })
    revalidatePath('/readings')
    return { ok: true, message: 'Reading created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create reading' } as const
  }
}
