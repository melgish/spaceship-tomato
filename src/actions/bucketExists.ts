'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to validate that a bucket exists in the database.
 * @param id Id of bucket to search for.
 * @returns true if bucket exists, otherwise false.
 */
export async function bucketExists(id: number = 0) {
  return id > 0 && (await prisma.bucket.count({ where: { id } })) > 0
}
