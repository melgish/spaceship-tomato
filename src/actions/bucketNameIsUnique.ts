'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to validate that the name of a bucket is unique.
 * @param name Name of the bucket to test.
 * @param id Optional id of bucket to exclude when editing an existing bucket.
 * @returns true if the name exists on another bucket, otherwise false.
 */
export async function bucketNameIsUnique(name: string, id: number = 0) {
  const where = { name, id: (id > 0 && { not: id }) || undefined }
  return (await prisma.bucket.count({ where })) === 0
}
