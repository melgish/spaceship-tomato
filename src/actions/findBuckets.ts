'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to fetch all of buckets from the database.
 * @returns An array of buckets. The returned data will include plant info if
 * the bucket contains a plant.
 */
export async function findBuckets() {
  return await prisma.bucket.findMany({
    include: { plant: true },
    orderBy: [{ name: 'asc' }],
  })
}

export type Buckets = Awaited<ReturnType<typeof findBuckets>>
export type Bucket = Buckets[number]
