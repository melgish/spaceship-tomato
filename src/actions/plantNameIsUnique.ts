'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to validate that the name of a plant is unique.
 * @param name Name of the plant to test.
 * @param id Optional id of plant to exclude when editing an existing plant.
 * @returns true if the name exists on another plant, otherwise false.
 */
export async function plantNameIsUnique(name: string, id: number = 0) {
  const where = { name, id: (id > 0 && { not: id }) || undefined }
  return (await prisma.plant.count({ where })) === 0
}
