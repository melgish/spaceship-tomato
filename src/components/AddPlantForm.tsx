'use client'
import { useId } from 'react'
import { useForm } from 'react-hook-form'
import { zodResolver } from '@hookform/resolvers/zod'
import ErrorMessage from '@/components/ErrorMessage'
import ResetButton from '@/components/ResetButton'
import SubmitButton from '@/components/SubmitButton'

import { addPlant } from '@/actions/addPlant'
import { type AddPlant, addPlantSchema } from '@/schemas/addPlant.schema'
import { notify } from '@/lib/notify'

const defaultValues: AddPlant = { name: '', minPH: 6.0, maxPH: 7.0 }
const resolver = zodResolver(addPlantSchema)

export default function AddPlantForm() {
  const id = useId()

  const clear = () => {
    reset(defaultValues)
  }

  const save = async (data: AddPlant) => {
    if (notify(await addPlant(data))) {
      clear()
    }
  }

  const { handleSubmit, formState, register, reset } = useForm({
    defaultValues,
    resolver,
    shouldFocusError: true,
  })

  return (
    <form className="flex gap-x-3" noValidate onSubmit={handleSubmit(save)}>
      <div className="required field">
        <label htmlFor={`${id}-name`}>Name</label>
        <input
          {...register('name')}
          id={`${id}-name`}
          maxLength={48}
          placeholder="Ex. Tomato"
          required
          size={32}
          type="text"
        />
        <ErrorMessage error={formState.errors?.name} />
      </div>
      <div className="required field">
        <label htmlFor={`${id}-minPH`}>Minimum PH</label>
        <input
          {...register('minPH')}
          id={`${id}-minPH`}
          max={14}
          min={0}
          placeholder="Ex. 7.0"
          required
          size={10}
          step=".1"
          type="number"
        />
        <ErrorMessage error={formState.errors?.minPH} />
      </div>
      <div className="required field">
        <label htmlFor={`${id}-maxPH`}>Maximum PH</label>
        <input
          {...register('maxPH')}
          id={`${id}-maxPH`}
          max={14}
          min={0}
          placeholder="Ex. 7.0"
          required
          size={10}
          step=".1"
          type="number"
        />
        <ErrorMessage error={formState.errors?.maxPH} />
      </div>
      <div className="flex items-center gap-x-2">
        <SubmitButton />
        <ResetButton onClick={clear} />
      </div>
    </form>
  )
}
