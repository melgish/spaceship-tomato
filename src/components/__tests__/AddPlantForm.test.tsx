import { Mock, beforeEach, describe, expect, it, vi } from 'vitest'
import { act, fireEvent, render, screen } from '@testing-library/react'
import AddPlantForm from '../AddPlantForm'
import { addPlant } from '@/actions/addPlant'
import { plantNameIsUnique } from '@/actions/plantNameIsUnique'

vi.mock('@/actions/plantNameIsUnique')
vi.mock('@/actions/addPlant')
vi.mock('react-hot-toast')

beforeEach(() => {
  // Mock deep validation
  ;(plantNameIsUnique as Mock).mockResolvedValue(true)
  ;(addPlant as Mock).mockResolvedValue({ ok: true, message: 'Good' })
})

it('should render', () => {
  const { container } = render(<AddPlantForm />)

  expect(container).toMatchSnapshot()
})

describe('when invalid and submit is clicked', () => {
  it('should report validation errors', async () => {
    const { container } = render(<AddPlantForm />)

    await act(() => fireEvent.click(screen.getByRole('button', { name: 'Submit' })))

    expect(container).toMatchSnapshot()
    expect(addPlant).not.toBeCalled()
  })
})

describe('when valid and submit is clicked', () => {
  it('should submit data', async () => {
    const { container } = render(<AddPlantForm />)

    const name = screen.getByLabelText(/name/i)
    const setName = { target: { value: 'My Test Plant' } }
    await act(() => fireEvent.change(name, setName))

    const minPH = screen.getByLabelText(/Minimum PH/i)
    const setMin = { target: { value: '6.2', valueAsNumber: 6.2 } }
    await act(() => fireEvent.change(minPH, setMin))

    const maxPH = screen.getByLabelText(/Maximum PH/i)
    const setMax = { target: { value: '7.1', valueAsNumber: 7.1 } }
    await act(() => fireEvent.change(maxPH, setMax))

    const submit = screen.getByRole('button', { name: /submit/i })
    await act(() => fireEvent.click(submit))

    expect(container).toMatchSnapshot()
    expect(addPlant).toBeCalledWith({
      name: 'My Test Plant',
      minPH: 6.2,
      maxPH: 7.1,
    })
  })
})
