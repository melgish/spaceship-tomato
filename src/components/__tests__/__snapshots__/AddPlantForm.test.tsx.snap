// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render 1`] = `
<div>
  <form
    class="flex gap-x-3"
    novalidate=""
  >
    <div
      class="required field"
    >
      <label
        for=":r0:-name"
      >
        Name
      </label>
      <input
        id=":r0:-name"
        maxlength="48"
        name="name"
        placeholder="Ex. Tomato"
        required=""
        size="32"
        type="text"
      />
    </div>
    <div
      class="required field"
    >
      <label
        for=":r0:-minPH"
      >
        Minimum PH
      </label>
      <input
        id=":r0:-minPH"
        max="14"
        min="0"
        name="minPH"
        placeholder="Ex. 7.0"
        required=""
        size="10"
        step=".1"
        type="number"
      />
    </div>
    <div
      class="required field"
    >
      <label
        for=":r0:-maxPH"
      >
        Maximum PH
      </label>
      <input
        id=":r0:-maxPH"
        max="14"
        min="0"
        name="maxPH"
        placeholder="Ex. 7.0"
        required=""
        size="10"
        step=".1"
        type="number"
      />
    </div>
    <div
      class="flex items-center gap-x-2"
    >
      <button
        class="emerald"
        type="submit"
      >
        Submit
      </button>
      <button
        class="slate"
        type="button"
      >
        Clear
      </button>
    </div>
  </form>
</div>
`;

exports[`when invalid and submit is clicked > should report validation errors 1`] = `
<div>
  <form
    class="flex gap-x-3"
    novalidate=""
  >
    <div
      class="required field"
    >
      <label
        for=":r1:-name"
      >
        Name
      </label>
      <input
        id=":r1:-name"
        maxlength="48"
        name="name"
        placeholder="Ex. Tomato"
        required=""
        size="32"
        type="text"
      />
      <div
        class="error text-rose-800"
      >
        Name is required
      </div>
    </div>
    <div
      class="required field"
    >
      <label
        for=":r1:-minPH"
      >
        Minimum PH
      </label>
      <input
        id=":r1:-minPH"
        max="14"
        min="0"
        name="minPH"
        placeholder="Ex. 7.0"
        required=""
        size="10"
        step=".1"
        type="number"
      />
    </div>
    <div
      class="required field"
    >
      <label
        for=":r1:-maxPH"
      >
        Maximum PH
      </label>
      <input
        id=":r1:-maxPH"
        max="14"
        min="0"
        name="maxPH"
        placeholder="Ex. 7.0"
        required=""
        size="10"
        step=".1"
        type="number"
      />
    </div>
    <div
      class="flex items-center gap-x-2"
    >
      <button
        class="emerald"
        type="submit"
      >
        Submit
      </button>
      <button
        class="slate"
        type="button"
      >
        Clear
      </button>
    </div>
  </form>
</div>
`;

exports[`when valid and submit is clicked > should submit data 1`] = `
<div>
  <form
    class="flex gap-x-3"
    novalidate=""
  >
    <div
      class="required field"
    >
      <label
        for=":r2:-name"
      >
        Name
      </label>
      <input
        id=":r2:-name"
        maxlength="48"
        name="name"
        placeholder="Ex. Tomato"
        required=""
        size="32"
        type="text"
      />
    </div>
    <div
      class="required field"
    >
      <label
        for=":r2:-minPH"
      >
        Minimum PH
      </label>
      <input
        id=":r2:-minPH"
        max="14"
        min="0"
        name="minPH"
        placeholder="Ex. 7.0"
        required=""
        size="10"
        step=".1"
        type="number"
      />
    </div>
    <div
      class="required field"
    >
      <label
        for=":r2:-maxPH"
      >
        Maximum PH
      </label>
      <input
        id=":r2:-maxPH"
        max="14"
        min="0"
        name="maxPH"
        placeholder="Ex. 7.0"
        required=""
        size="10"
        step=".1"
        type="number"
      />
    </div>
    <div
      class="flex items-center gap-x-2"
    >
      <button
        class="emerald"
        type="submit"
      >
        Submit
      </button>
      <button
        class="slate"
        type="button"
      >
        Clear
      </button>
    </div>
  </form>
</div>
`;
